File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    maxi=0
    def depth_dfs(self,node: Optional[TreeNode])-> int:
            if not node:
                return 0
            left=self.depth_dfs(node.left)
            right=self.depth_dfs(node.right)
            self.maxi=max(self.maxi,left+right)
            return max(left,right)+1
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.depth_dfs(root)
        return self.maxi

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max=0;
    public int diameterOfBinaryTree(TreeNode root) {
        depth_dfs(root);
        return max;
    }
    public int depth_dfs(TreeNode node){
        if(node==null){
            return 0;
        }
        int left=depth_dfs(node.left);
        int right=depth_dfs(node.right);
        max=Math.max(max,left+right);
        return Math.max(left,right)+1;
    }
}